File: solve/11725.py
# https://www.acmicpc.net/problem/11725

import sys
sys.setrecursionlimit(222_222)
n = int(sys.stdin.readline())
connections = [set() for i in range(n + 5)]
for i in range(n - 1):
    a, b = map(int, sys.stdin.readline().split())
    connections[a].add(b)
    connections[b].add(a)

parents = [0]*(n + 5)
def dfs(start, parent):
    parents[start] = parent
    for connected in connections[start]:
        if connected == parent:
            continue
        dfs(connected, start)

dfs(1, 0)
for i in range(2, n + 1):
    print(parents[i])

File: solve/11725.cpp
// https://www.acmicpc.net/problem/11725

#include <iostream>
#include <cstdio>
#include <algorithm>
#include <vector>

using namespace std;

int n;
vector<int> edges[111'111];
bool visited[111'111];
int parent[111'111];

void solve(int start)
{
    for (const auto& next : edges[start])
    {
        if (!visited[next])
        {
            visited[next] = true;
            parent[next] = start;
            solve(next);
        }
    }
}

int main() 
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    int n, start;
    cin >> n;
    for (int i = 0; i < n - 1; i++)
    {
        int u, v;
        cin >> u >> v;
        edges[u].push_back(v);
        edges[v].push_back(u);
    }
    solve(1);
    for (int i = 2; i <= n; i++)
    {
        cout << parent[i] << '\n';
    }
    return 0;
}